pkg/common/utils/gang_utils.go

```diff
@@ -0,0 +1,179 @@
+	// total length no longer than 20 + 28 + 5 + 10 = 63
+	shortTaskGroupName := fmt.Sprintf("%.20s", taskGroupName)
+	shortAppID := fmt.Sprintf("%.28s", appID)
+	return "tg-" + shortTaskGroupName + "-" + shortAppID + fmt.Sprintf("-%d", index)
+}
+
+func GetPlaceholderResourceRequest(resources map[string]resource.Quantity) v1.ResourceList {
+	resourceReq := v1.ResourceList{}
+	for k, v := range resources {
+		resourceReq[v1.ResourceName(k)] = v
+	}
+	return resourceReq
+}
+
+func GetPlaceholderFlagFromPodSpec(pod *v1.Pod) bool {
+	if value, ok := pod.Annotations[constants.AnnotationPlaceholderFlag]; ok {
+		if v, err := strconv.ParseBool(value); err == nil {
+			return v
+		}
+	}
+	return false
+}
+
+func GetTaskGroupFromPodSpec(pod *v1.Pod) string {
+	if value, ok := pod.Annotations[constants.AnnotationTaskGroupName]; ok {
+		return value
+	}
+	return ""
+}
+
+func GetTaskGroupsFromAnnotation(pod *v1.Pod) ([]v1alpha1.TaskGroup, error) {
+	taskGroupInfo, ok := pod.Annotations[constants.AnnotationTaskGroups]
+	if !ok {
+		return nil, nil
+	}
+	taskGroups := []v1alpha1.TaskGroup{}
+	err := json.Unmarshal([]byte(taskGroupInfo), &taskGroups)
+	if err != nil {
+		return nil, err
+	}
+	// json.Unmarchal won't return error if name or MinMember is empty, but will return error if MinResource is empty or error format.
+	for _, taskGroup := range taskGroups {
+		if taskGroup.Name == "" {
+			return nil, fmt.Errorf("can't get taskGroup Name from pod annotation, %s",
+				pod.Annotations[constants.AnnotationTaskGroups])
+		}
+		if taskGroup.MinMember == int32(0) {
+			return nil, fmt.Errorf("can't get taskGroup MinMember from pod annotation, %s",
+				pod.Annotations[constants.AnnotationTaskGroups])
+		}
+	}
+	return taskGroups, nil
+}
+
+type TaskGroupInstanceCountMap struct {
+	counts map[string]int32
+	sync.RWMutex
+}
+
+func NewTaskGroupInstanceCountMap() *TaskGroupInstanceCountMap {
+	return &TaskGroupInstanceCountMap{
+		counts: make(map[string]int32),
+	}
+}
+
+func (t *TaskGroupInstanceCountMap) Add(taskGroupName string, num int32) {
+	t.update(taskGroupName, num)
+}
+
+func (t *TaskGroupInstanceCountMap) AddOne(taskGroupName string) {
+	t.update(taskGroupName, 1)
+}
+
+func (t *TaskGroupInstanceCountMap) DeleteOne(taskGroupName string) {
+	t.update(taskGroupName, -1)
+}
+
+func (t *TaskGroupInstanceCountMap) update(taskGroupName string, delta int32) {
+	t.Lock()
+	defer t.Unlock()
+	if v, ok := t.counts[taskGroupName]; ok {
+		t.counts[taskGroupName] = v + delta
+	} else {
+		t.counts[taskGroupName] = delta
+	}
+}
+
+func (t *TaskGroupInstanceCountMap) Size() int {
+	t.RLock()
+	defer t.RUnlock()
+	return len(t.counts)
+}
+
+func (t *TaskGroupInstanceCountMap) GetTaskGroupInstanceCount(groupName string) int32 {
+	t.RLock()
+	defer t.RUnlock()
+	return t.counts[groupName]
+}
+
+func (t *TaskGroupInstanceCountMap) Equals(target *TaskGroupInstanceCountMap) bool {
+	t.RLock()
+	defer t.RUnlock()
+
+	if target == nil {
+		return false
+	}
```

Comment:
I think we need to add the cases when t is nil and when both t and target is nil.

Reply:
Good suggestion. I just realize this is something different in golang
https://tour.golang.org/methods/12
I will add the check and the UT for this in the next commit.